ethcore/src/trace/db.rs

```diff
@@ -0,0 +1,573 @@
+		tracedb_path.push("tracedb");
+		let tracesdb = Database::open_default(tracedb_path.to_str().unwrap()).unwrap();
+
+		// check if in previously tracing was enabled
+		let tracing_was_enabled = match tracesdb.get(b"enabled").unwrap() {
+			Some(ref value) if value as &[u8] == &[0x1] => Some(true),
+			Some(ref value) if value as &[u8] == &[0x0] => Some(false),
+			Some(_) => { panic!("tracesdb is malformed") },
+			None => None,
+		};
+
+		// compare it with the current option.
+		let tracing = match (tracing_was_enabled, config.enabled) {
+			(Some(true), Some(true)) => true,
+			(Some(true), None) => true,
+			(Some(true), Some(false)) => false,
+			(Some(false), Some(true)) => { panic!("Tracing can't be enabled. Resync required."); },
+			(Some(false), None) => false,
+			(Some(false), Some(false)) => false,
+			(None, Some(true)) => true,
+			_ => false,
+		};
+
+		config.enabled = Some(tracing);
+
+		let encoded_tracing= match tracing {
+			true => [0x1],
+			false => [0x0]
+		};
+
+		tracesdb.put(b"enabled", &encoded_tracing).unwrap();
+
+		Tracedb {
+			traces: RwLock::new(HashMap::new()),
+			blooms: RwLock::new(HashMap::new()),
+			tracesdb: tracesdb,
+			config: config,
+			extras: extras,
+		}
+	}
+
+	/// Returns traces for block with hash.
+	fn traces(&self, block_hash: &H256) -> Option<BlockTraces> {
+		self.tracesdb.read_with_cache(&self.traces, block_hash)
+	}
+
+	fn matching_block_traces(
+		&self,
+		filter: &Filter,
+		traces: FlatBlockTraces,
+		block_hash: H256,
+		block_number: BlockNumber
+	) -> Vec<LocalizedTrace> {
+		let tx_traces: Vec<FlatTransactionTraces> = traces.into();
+		tx_traces.into_iter()
+			.enumerate()
+			.flat_map(|(tx_number, tx_trace)| {
+				self.matching_transaction_traces(filter, tx_trace, block_hash.clone(), block_number, tx_number)
+			})
+			.collect()
+	}
+
+	fn matching_transaction_traces(
+		&self,
+		filter: &Filter,
+		traces: FlatTransactionTraces,
+		block_hash: H256,
+		block_number: BlockNumber,
+		tx_number: usize
+	) -> Vec<LocalizedTrace> {
+		let tx_hash = self.extras.transaction_hash(block_number, tx_number)
+		.expect("Expected to find transaction hash. Database is probably malformed");
+
+		let flat_traces: Vec<FlatTrace> = traces.into();
+		flat_traces.into_iter()
+			.enumerate()
+			.filter_map(|(index, trace)| {
+				match filter.matches(&trace) {
+					true => Some(LocalizedTrace {
+						parent: trace.parent,
+						children: trace.children,
+						depth: trace.depth,
+						action: trace.action,
+						result: trace.result,
+						trace_number: index,
+						transaction_number: tx_number,
+						transaction_hash: tx_hash.clone(),
+						block_number: block_number,
+						block_hash: block_hash
+					}),
+					false => None
+				}
+			})
+			.collect()
+	}
+}
+
+impl<T> TraceDatabase for Tracedb<T> where T: DatabaseExtras {
+	fn tracing_enabled(&self) -> bool {
+		self.config.enabled.expect("Auto tracing hasn't been properly configured.")
```

Comment:
What do you mean exactly?

The only way to avoid `unwrapping` option here is to store `enabled` field as a separate property. Do you think it is better?